File: 백준/Gold/1987. 알파벳/알파벳.cc
#include<iostream>
using namespace std;
int r,c,ans;
string s;
bool inrange(int x, int y){
    return x >=1 && x <=r && y >=1 && y <= c;
}
int check[30];
int dx[4] = {1,-1,0,0};
int dy[4] = {0,0,1,-1};
char arr[22][22];
int visit[22][22];

void go(int x, int y){
    int tmp = 0;
    for(int i = 0; i<30; i++){
        if(check[i]) tmp++;
    }
    ans = max(ans, tmp);
    
    for(int i = 0; i<4; i++){
        int nx = x + dx[i];
        int ny = y + dy[i];
        int nxt = arr[nx][ny] - 'A';//다음 알파벳
        if(!inrange(nx,ny) || check[nxt] || visit[nx][ny]) continue;
        check[nxt] = 1;
        visit[nx][ny] = 1;
        go(nx,ny);
        check[nxt] = 0;
        visit[nx][ny] = 0;
    }
}

int main()
{
    ios::sync_with_stdio(0); cin.tie(0);
    cin >> r >> c;
    for(int i = 1; i<=r; i++){
        cin >> s;
        for(int j = 1; j<=c; j++) arr[i][j] = s[j-1];
    }
    visit[1][1] = 1;
    check[arr[1][1]-'A'] = 1;
    go(1,1);
    cout << ans;
    return 0;
}

File: 백준/Gold/1987. 알파벳/알파벳.java
import java.io.*;
import java.util.*;

class Main {
	static int R,C,ans;
	static String[][] arr;
	static int[] dx = {-1,1,0,0};
	static int[] dy = {0,0,-1,1};
	static boolean[][] visited;
	public static void main(String[] args) throws Exception {
		BufferedReader br = new BufferedReader(new InputStreamReader(System.in));
		StringTokenizer st = new StringTokenizer(br.readLine());
		R = Integer.parseInt(st.nextToken());
		C = Integer.parseInt(st.nextToken());
		arr = new String[R][C];
		visited = new boolean[R][C];
		
		for(int i = 0; i < R; i++){
			String input = br.readLine();
			for(int j = 0; j < C; j++) arr[i][j] = input.charAt(j)+"";
		}
		go(0,0,arr[0][0]);
		visited[0][0] = true;
		System.out.println(ans);
	}

	static void go(int cx, int cy, String tmp){
		ans = Math.max(tmp.length(), ans);

		for(int i = 0; i < 4; i++){
			int nx = cx + dx[i];
			int ny = cy + dy[i];
			if(nx < 0 || nx >= R || ny < 0 || ny >= C) continue;
			if(tmp.contains(arr[nx][ny])) continue;
			if(visited[nx][ny]) continue;
			visited[nx][ny] = true;
			go(nx,ny,tmp + arr[nx][ny]);
			visited[nx][ny] = false;
		}
	}
}